### ISSUE TITLE
ETag not set correctly

### ISSUE THREAD

**Issue by richorama** [OrleansContrib/Orleans.StorageProvider.Redis#9]
A Guid is just made up on reads: 

https://github.com/OrleansContrib/Orleans.StorageProvider.Redis/blob/dev/RedisStorage/RedisStorage.cs#L124

**Comment by richorama** [comment]
Redis does not seem to support a native ETag feature. At the moment the ETag is a random guid generated on read (bad).

Perhaps when the provider writes to storage, it should wrap the object with another object to add an ETag:

```cs
class WithETag
{
  public object Value {get;set;}
  public string ETag {get;set;}
}
```

It can then decompose this when reading.

Backward compatibility can be maintained if you test the object type when reading from storage.

**Comment by richorama** [comment]
@pditterline curious to know what your thoughts are on this?

**Comment by pditterline** [comment]
Hmm, that's definitely a problem. We discussed this here and I think the options are to either wrap the object like you mention above, or to insert a second key with _etag appended (or something similar) that has the etag as a value. We have a similar problem with an internal Redis membership provider that we implemented.

The issue will be ensuring transactional integrity on the read/write, but we should be able to do that with either solution I'd think. I'm not a Redis expert by any means so I don't know what the best answer is.
